### ISSUE TITLE
Moving rows_entry is expensive for large partitions

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#3247]
Moving a single `rows_entry` is O(log(number of rows in partition)), because `boost::intrusive::rbtree_algorithms::replace_node()` has to locate the header of the tree. This causes latency spikes during LSA segment compaction when there are partitions with many rows.

```
(inlined by) boost::intrusive::compact_rbtree_node_traits_impl<void*>::get_parent(boost::intrusive::compact_rbtree_node<void*>* const&) at /usr/include/boost/intrusive/detail/rbtree_node.hpp:137
 (inlined by) boost::intrusive::bstree_algorithms_base<boost::intrusive::rbtree_node_traits<void*, true> >::get_header(boost::intrusive::compact_rbtree_node<void*> const* const&) at /usr/include/boost/intrusive/detail/bstree_algorithms_base.hpp:168
 (inlined by) boost::intrusive::rbtree_algorithms<boost::intrusive::rbtree_node_traits<void*, true> >::replace_node(boost::intrusive::compact_rbtree_node<void*>* const&, boost::intrusive::compact_rbtree_node<void*>* const&) at /usr/include/boost/intrusive/rbtree_algorithms.hpp:234
 (inlined by) intrusive_set_external_comparator_member_hook::intrusive_set_external_comparator_member_hook(intrusive_set_external_comparator_member_hook&&) at /home/tgrabiec/src/scylla/intrusive_set_external_comparator.hh:43
 (inlined by) rows_entry::rows_entry(rows_entry&&) at /home/tgrabiec/src/scylla/mutation_partition.cc:1370
standard_migrator<rows_entry>::migrate(void*, void*) const at /home/tgrabiec/src/scylla/./utils/allocation_strategy.hh:67
logalloc::region_impl::compact(logalloc::segment*, logalloc::segment_descriptor&)::{lambda(logalloc::region_impl::object_descriptor const*, void*)#1}::operator()(logalloc::region_impl::object_descriptor const*, void*) const at /home/tgrabiec/src/scylla/utils/logalloc.cc:1193
 (inlined by) void logalloc::region_impl::for_each_live<logalloc::region_impl::compact(logalloc::segment*, logalloc::segment_descriptor&)::{lambda(logalloc::region_impl::object_descriptor const*, void*)#1}>(logalloc::segment*, logalloc::region_impl::compact(logalloc::segment*, logalloc::segment_descriptor&)::{lambda(logalloc::region_impl::object_descriptor const*, void*)#1}&&) at /home/tgrabiec/src/scylla/utils/logalloc.cc:1138
 (inlined by) logalloc::region_impl::compact(logalloc::segment*, logalloc::segment_descriptor&) at /home/tgrabiec/src/scylla/utils/logalloc.cc:1190
 (inlined by) logalloc::region_impl::compact_single_segment_locked() at /home/tgrabiec/src/scylla/utils/logalloc.cc:1439
logalloc::region_impl::compact() at /home/tgrabiec/src/scylla/utils/logalloc.cc:1446
```

**Comment by avikivity** [comment]
Interesting problem. How do you plan to fix it?

Moving to a btree would, but it's a lot of work. Add a pointer to the root?

**Comment by linux-wizard** [comment]
Hi, @tgrabiec , can I know what is the status for this performance issue? Has been solved or do we consider this as a ScyllaDB anti-pattern?

**Comment by denesb** [comment]
Overly large partitions are considered an anti-pattern in Scylla. That said we are actively working on lessening the impact of large partition on Scylla. Even though it is an anti-pattern, large partitions are a fact of life, partly because some users don't know they are bad, partly because large partitions can form without the user knowing it (by accumulating a lot of rows over time).

As for this particular issue, it is not solved yet.

**Comment by linux-wizard** [comment]
What is considered a large partition? Partition size (our partitions are already shared to keep Max row size below 1MB) or number of rows (we may have up to 1000 rows) or a combination of both?

**Comment by avikivity** [comment]
For this problem, the number of rows is the problem. If you have 1000 rows in a partition, and all are cached, then to evict or move one of them we incur 10 memory accesses on average. If all of them are uncached, that's about 1 microsecond (in total) that the CPU spends per row.

If the rows are also small, then we didn't get much benefit from moving or evicting them. So this issue is problematic with large numbers of very small rows (100 bytes or less). Large rows actually make this problem less severe, because if rows are large, we have to move or evict a smaller number of them.

**Comment by slivne** [comment]
@xemul is working on moving cache and memtables to B+tree - this should be revaluated one this done and hopefully avi is correct (https://github.com/scylladb/scylla/issues/3247#issuecomment-369614963) and it "will be gone"

**Comment by xemul** [comment]
For the record. The patchset "switch partitions cache from BST to B+" addresses similar issue, but with cache_entry/memtable_entry -- eviction is logN there with std::set and is constant with B+.

**Comment by xemul** [comment]
Also, not only moving is expensive, but also eviction (which implies erasure by node iterator) is logN as it also walks up the root of the in destructor.

**Comment by xemul** [comment]
@tgrabiec , the merged 873e7320 should be the fix for this issue. Moving a `rows_entry` is now constant time.
